src/gpt.c

```diff
@@ -0,0 +1,346 @@
+	struct fdisk_table *tb = NULL;
+	struct fdisk_iter *itr = NULL;
+	struct fdisk_partition *pa = NULL;
+	fdisk_sector_t number_of_sectors;
+	fdisk_sector_t region_start_sector, region_end_sector;
+
+	g_return_val_if_fail(cxt, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	grain_size = fdisk_get_grain_size(cxt);
+	sector_size = fdisk_get_sector_size(cxt);
+
+	if (region_start < 34*512 || region_size == 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region configuration is invalid");
+		goto out;
+	}
+
+	if ((region_start % grain_size) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region start %"G_GINT64_MODIFIER "d is not aligned to grain-size %"G_GINT64_MODIFIER "d",
+				region_start, 2 * grain_size);
+		goto out;
+	}
+
+	if ((region_size % (2 * grain_size)) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region size %"G_GINT64_MODIFIER "d is not aligned to the double grain-size %"G_GINT64_MODIFIER "d",
+				region_size, 2 * grain_size);
+		goto out;
+	}
+
+	number_of_sectors = fdisk_get_nsectors(cxt);
+	if ((region_start + region_size) / sector_size >= number_of_sectors) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region configuration is bigger than device");
+		res = FALSE;
+		goto out;
+	}
+
+	sector_size = fdisk_get_sector_size(cxt);
+	region_start_sector = region_start / sector_size;
+	region_end_sector = region_start_sector + region_size / sector_size - 1;
+
+	if (fdisk_get_partitions(cxt, &tb)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Failed to get partitions");
+		return FALSE;
+	}
+
+	itr = fdisk_new_iter(FDISK_ITER_FORWARD);
+	if (!itr)
+		g_error("%s: Failed to allocate libfdisk iter\n", G_STRLOC);
+	while (fdisk_table_next_partition(tb, itr, &pa) == 0) {
+		fdisk_sector_t p_start_sector, p_end_sector;
+		if (!fdisk_partition_has_start(pa) ||
+		    !fdisk_partition_has_end(pa) ||
+		    !fdisk_partition_has_partno(pa))
+			g_error("%s: Invalid partition entry\n", G_STRLOC);
+
+		/* skip boot partition entry */
+		if (fdisk_partition_get_partno(pa) == BOOT_PARTITION_ENTRY) {
+			found = TRUE;
+			continue;
+		}
+
+		p_start_sector = fdisk_partition_get_start(pa);
+		p_end_sector = fdisk_partition_get_end(pa);
+
+		if (region_start_sector <= p_end_sector && p_start_sector <= region_end_sector) {
+			g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+					"Region (sectors 0x%"G_GINT64_MODIFIER "x - 0x%"G_GINT64_MODIFIER "x) overlaps"
+					"with partition %zd (sectors 0x%"G_GINT64_MODIFIER "x - 0x%"G_GINT64_MODIFIER "x)",
+					region_start_sector, region_end_sector,
+					fdisk_partition_get_partno(pa),
+					p_start_sector, p_end_sector);
+			goto out_table;
+		}
+	}
+
+	if (!found) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"No boot partition found in entry %d",
+				BOOT_PARTITION_ENTRY);
+		goto out_table;
+	}
```

Comment:
The check for the existence of a boot partition is done in lines +226 to +231 of `r_gpt_switch_get_inactive_partition()` and lines +307 to +312 of `r_gpt_switch_set_boot_partition()` as well. Why have it here in a different form? Are there cases that we are not going to catch if we delete this check here?

Reply:
We reload the partition table, so I wanted to be extra sure.

Reply:
Yes, that sounds reasonable.